src/Test.UnitTests.Sarif/RoundTrippingTests.cs

```diff
@@ -0,0 +1,102 @@
+// license. See LICENSE file in the project root for full license information. 
+
+using System;
+using System.Globalization;
+using FluentAssertions;
+using Newtonsoft.Json;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Sarif
+{
+    public class RoundTrippingTests
+    {
+        private static ResourceExtractor s_extractor = new ResourceExtractor(typeof(RoundTrippingTests));
+
+        private static string GetResourceContents(string resourceName)
+            => s_extractor.GetResourceText($"RoundTripping.{resourceName}");
+
+        // Class used to test serialization and deserialization of an arbitrary class.
+        public class CustomObject
+        {
+            public DateTime DateTime { get; set; }
+            public string String { get; set; }
+        }
+
+        private static DateTime ParseUtcDateTime(string dateTime)
+        {
+            // DateTimeStyles.AdjustToUniversal is necessary because otherwise DateTime.Parse
+            // adjusts the time from UTC to local time (for example, in Redmond, it subtracts
+            // 7 hours). We want the hours field to be exactly as specified in the test file.
+            return DateTime.Parse(dateTime, provider: null, DateTimeStyles.AdjustToUniversal);
+        }
+
+        [Fact]
+        public void SarifLog_PropertyBagProperties_CanBeRoundTripped()
+        {
+            string originalContents = GetResourceContents("RoundTripping.sarif");
+
+            SarifLog log = SarifUtilities.DeserializeObject<SarifLog>(originalContents);
+
+            PropertyBagHolder holder = log.Runs[0].Results[0];
+
+            DateTime dateTimeProperty = holder.GetProperty<DateTime>("dateTime");
+            dateTimeProperty.Should().Be(ParseUtcDateTime("2019-09-25T15:07:01Z"));
+
+            string stringProperty = holder.GetProperty("string");
+            stringProperty.Should().Be("Hello, string!");
+
+            int intProperty = holder.GetProperty<int>("int");
+            intProperty.Should().Be(42);
+
+            long longProperty = holder.GetProperty<long>("long");
+            longProperty.Should().Be(5000000000);
+
+            // This integer-valued JSON property is too big for an Int32.
+            Action action = () => holder.GetProperty<int>("long");
+            action.Should().Throw<JsonReaderException>();
+
+            bool trueProperty = holder.GetProperty<bool>("true");
+            trueProperty.Should().BeTrue();
+
+            bool falseProperty = holder.GetProperty<bool>("false");
+            falseProperty.Should().BeFalse();
+
+            string nullProperty = holder.GetProperty("null");
+            nullProperty.Should().BeNull();
+
+            DateTime[] dateTimeArray = holder.GetProperty<DateTime[]>("dateTimeArray");
+            dateTimeArray.Length.Should().Be(2);
+            dateTimeArray[0].Should().Be(ParseUtcDateTime("2019-09-26T15:52:02Z"));
+            dateTimeArray[1].Should().Be(ParseUtcDateTime("2019-09-26T15:54Z"));
```

Comment:
How much precision do we emit for DateTimes? Does UtcNow roundtrip properly?

Reply:
Does `UtcNow` round-trip properly? Yes.

How much precision do we emit for `DateTime`s?

    - For `DateTime`s in "first-class" properties, we emit three decimal digits. Example: `log.Runs[0].Invocations[0].StartTimeUtc = DateTime.UtcNow;` => `"startTimeUtc": "2019-09-30T17:10:20.185Z"`. 
    - For `DateTime`s in property bag properties, we emit seven decimal digits. Example: `log.Runs[0].SetProperty("now", DateTime.UtcNow)` => `"properties": { "now": "2019-09-30T17:10:20.1851531Z" }` .

The reason for the difference is subtle. First-class `DateTime`-valued properties are decorated with `[JsonConvert(typeof(DateTimeConverter)]`, and `DateTimeConverter` uses `SarifUtilities.SarifDateTimeFormatMillisecondsPrecision`. But `PropertyBagHolder.SetProperty` calls `JsonConvert.SerializeObject` _without a contract resolver_, so the value is serialized with Newtonsoft.Json's default `DateTime` serialization, which (apparently) uses more precision.

`SetProperty` doesn't set a contract resolver because it doesn't know whether it's working with a v1 or a v2 log, so it doesn't know whether to use the v1 or v2 contract resolver. However, I can add code to use the `DateTimeConverter` for `DateTime`-valued properties. Then everything will serialize with millisecond precision.

---
In reply to: [329664303](https://github.com/microsoft/sarif-sdk/pull/1688#discussion_r329664303) [](ancestors = 329664303)

Reply:
Ugh. Doing that breaks the unit test `PropertyBagHolder_SetProperty_SetsDateTimeProperty`, which has the comment:

> DateTime instances should not be converted to UTC when persisted to a property bag.  The reason is that we can't be  sure that the user has an actual UTC time in hand, or even the ability to convert the property date time to UTC.

Nightmare. I need to think more about this. After lunch.

---
In reply to: [329717572](https://github.com/microsoft/sarif-sdk/pull/1688#discussion_r329717572) [](ancestors = 329717572,329664303)

Reply:
Ok, I've thought about it. I can live with `DateTime`-valued property bag properties _that are created with `SetProperty`_ having more digits than first-class `DateTime`-valued properties. It's not great but IMO we're at the point of diminishing returns on this.

---
In reply to: [329725369](https://github.com/microsoft/sarif-sdk/pull/1688#discussion_r329725369) [](ancestors = 329725369,329717572,329664303)